api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ValueSetTranslatorImpl.java

```diff
@@ -72,45 +70,52 @@ public ValueSet toFhirResource(@Nonnull Concept concept) {
 		
 		ValueSet.ValueSetComposeComponent compose = new ValueSet.ValueSetComposeComponent();
 		
+		List<ValueSet.ConceptSetComponent> sets = new ArrayList<ValueSet.ConceptSetComponent>();
+		ValueSet.ConceptSetComponent conceptUuidSet = new ValueSet.ConceptSetComponent();
+		sets.add(conceptUuidSet);
+		
+		Map<ConceptSource, String> conceptSourceCache = new HashMap<>();
+		
 		for (ConceptSet conceptSet : concept.getConceptSets()) {
 			Concept conceptSetMember = conceptSet.getConcept();
 			
 			if (conceptSetMember != null) {
-				if (conceptSetMember.getSet()) {
-					ValueSet.ConceptSetComponent conceptSetComponent = new ValueSet.ConceptSetComponent();
-					conceptSetComponent
-					        .addValueSet(valueSetReferenceTranslator.toFhirResource(conceptSetMember).getReference());
-					compose.addInclude(conceptSetComponent);
-					continue;
-				}
+				// first, add the uuid and display to the concept uuid set
+				ValueSet.ConceptReferenceComponent uuidConceptReferenceComponent = new ValueSet.ConceptReferenceComponent();
```

Comment:
So a point that occurs to me here: in general, in FHIR, we wouldn't include a ValueSet as part of a ValueSet directly. FHIR distinguishes between [intensional and extensional ValueSets](https://www.hl7.org/fhir/valueset.html#int-ext), i.e., those value sets that are defined as "include the terms from Concept Source A and Value Set B" and those that are just an enumeration of all the concepts contained in the set. The implementation here was supposed to just represent extensional value sets (the ones where we enumerate all the concepts in the set).

Because FHIR ValueSets are not (in FHIR terms) concepts, the ValueSet itself would never appear in the extension (or "expansion"—FHIR's term for the operation that converts an intensional ValueSet into it's extensional representation). Thus, in the original implementation, we skip any set members that are themselves sets, instead just including their set members directly.

Obviously, in OpenMRS-world, things are different, i.e., I suppose it's possible for a concept that defines a concept set to also have some meaning other than just defining a set of additionally included concepts (i.e., we might use it to define a concept hierarchy).

I'm not really sure what the right thing to do here is, but I'm a little hesitant to directly include those concept sets here. Maybe what we could do is use the `ValueSet.compose.include.valueSet` parameter to reference the included concept set?

Reply:
@ibacher ah, so maybe I misunderstood what the original code was supposed to be doing... are you saying that it's supposed to flatten it out... ie, if it finds a member that is in itself a set, retrieve all the set members and just add them to the top-level list (and then proceeding recursively if there are set members that are sets and in turn have sets?) while skipping the set concept?  I'm fine with that.

When I saw the getReference() call in the old code, it looks like it was building some sort of reference, and I didn't know how to handle that in my new model.  Maybe it was already doing what you suggested in your last paragraph?  If so, I can add some tests around it to confirm.

Reply:
Oh, looking closer, I do think it was doing what you suggested in the last paragraph... let me see if I can restore that and add some more tests and update the PR

Reply:
Yes, I think that's what it was doing previously, but I don't think we had any tests to cover it.